### ISSUE TITLE
Don't append but insert into sys path

### ISSUE THREAD

**Issue by Solomon1732** [Tribler/py-ipv8#642]
#### Platform #####
All
#### Expected behavior ####
Append paths to system path in `setup.py`.
#### Actual behavior #####
System path is being set up in various modules in the code, resulting in duplicates and a setup that's hard to keep track of.
https://github.com/Tribler/py-ipv8/search?q=sys.path.append&unscoped_q=sys.path.append

An example of a duplicate is `twisted/plugins/ipv8_plugin.py` and `twisted/plugins/trustchain_crawler_plugin.py`, both of which add `twisted` to the system path. Is it possible to gather them all up in setup.py?

As a minor note, appending to the `sys path` means that "an installed package will take precedence over the local package in development (and hair pulling may ensue)." (as the comment to [this](https://stackoverflow.com/questions/1893598/pythonpath-vs-sys-path/1893663#1893663) stackoverflow points out. `sys.path.insert(1,` can overcome it.

**Comment by Solomon1732** [comment]
Another thing I noticed (thanks to [this](https://stackoverflow.com/questions/1893598/pythonpath-vs-sys-path/1893805#1893805)) is that hardcoding something like
```python
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..')))
```
is brittle since it assumes the directory structure won't change.

I took the example from here https://github.com/Tribler/py-ipv8/blob/702e5309c36602d72d65bd6af271588d819b278e/twisted/plugins/trustchain_crawler_plugin.py#L14

**Comment by qstokkink** [comment]
> Is it possible to gather them all up in setup.py?

No. We assume the scripts can be run from any working directory. If we want to import anything from `setup.py`, we will have to insert to the python path to import from that file (which was the problem to solve in the first place). Any script running stand-alone will need this (or we force everyone to do system environment `PYTHONPATH` modification which leads to even more hair pulling).

> As a minor note, appending to the sys path means that "an installed package will take precedence over the local package in development (and hair pulling may ensue)." (as the comment to this stackoverflow points out. sys.path.insert(1, can overcome it.

I agree. This would be a good change. In the past IPv8 was never on `pip`, so there were no installed off-site packages. Now that we are installable with a `setup.py` we should address this.

**Comment by Solomon1732** [comment]
If I were to PR this, does it go to the `asyncio` branch?

**Comment by qstokkink** [comment]
We probably don't need to backport this to master, just `asyncio` should be fine.

**Comment by Solomon1732** [comment]
Is it possible to perhaps move it to the `__init__.py` in each directory?

**Comment by qstokkink** [comment]
Hmm.. it would be nice if that worked. Let me test that.

**Comment by qstokkink** [comment]
First test: definitely can't use `__init__.py` for importing with Twisted plugins.

Backporting is not a high priority though, let me see if the new `asyncio` structure allows this.

**Comment by qstokkink** [comment]
Test 2: Works, but not entirely as expected 😃 

Given an `__init__.py` file in the `scripts` folder:

```python
import os
import sys

sys.path.insert(1, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
```

Any script would have to `import __init__` manually before the `sys.path` is modified. Ex:

```python
import __init__

import something

print(something.HELLO)
```

This is a bit.. ugly. For clarity the `__init__.py` should probably be renamed to `__scriptpath__.py` , or something. At least the underscores in the naming guarantee that the linters will not go bananas over the import order.

**Comment by Solomon1732** [comment]
It is ugly, I agree. Which direction do you think is best to take with such a PR? Either using `insert` in-place or using something like `__scriptpath__,py` could work. The latter one doesn't have to replace an `__init__.py`, merely be along side it, no?

**Comment by qstokkink** [comment]
> The latter one doesn't have to replace an __init__.py, merely be along side it, no?

Indeed (though technically there is no `__init__.py` to replace in the `scripts` folder).

I do very much like the fact of only doing ugly path modification once though.

For now I think an `import __scriptpath__` is the best option. At least it's a step up from how we had to do it before with Twisted.

**Comment by Solomon1732** [comment]
Will open a PR 👍

**Comment by qstokkink** [comment]
Thanks 👍

**Comment by Solomon1732** [comment]
Fixed by #649
